File: src/jira_dc_mcp/tools/issues.py
"""Issue lookup tools."""
from __future__ import annotations

import json

import httpx

from ..client import JiraClient


async def get_issue(client: JiraClient, issue_key: str, fields: str | None = None) -> str:
    """Get a Jira issue by key with selected fields."""
    try:
        issue = await client.get_issue(issue_key, fields=fields)

        result: dict = {
            "key": issue.get("key"),
            "id": issue.get("id"),
            "self": issue.get("self"),
        }

        f = issue.get("fields", {})
        result["fields"] = {
            "summary": f.get("summary"),
            "status": _name(f.get("status")),
            "issuetype": _name(f.get("issuetype")),
            "priority": _name(f.get("priority")),
            "resolution": _name(f.get("resolution")),
            "assignee": _user(f.get("assignee")),
            "reporter": _user(f.get("reporter")),
            "created": f.get("created"),
            "updated": f.get("updated"),
            "resolutiondate": f.get("resolutiondate"),
            "labels": f.get("labels", []),
            "components": [c.get("name") for c in f.get("components", [])],
            "description": f.get("description"),
        }

        # Include any custom fields that were explicitly requested
        if fields:
            requested = {fid.strip() for fid in fields.split(",")}
            for fid in requested:
                if fid.startswith("customfield_") and fid not in result["fields"]:
                    result["fields"][fid] = f.get(fid)

        # Include links
        links = f.get("issuelinks", [])
        if links:
            result["fields"]["issuelinks"] = [
                {
                    "type": link.get("type", {}).get("name"),
                    "direction": "outward" if "outwardIssue" in link else "inward",
                    "issue": (link.get("outwardIssue") or link.get("inwardIssue", {})).get("key"),
                    "summary": (link.get("outwardIssue") or link.get("inwardIssue", {})).get("fields", {}).get("summary"),
                }
                for link in links
            ]

        # Include comments count and last few
        comment_data = f.get("comment", {})
        if comment_data:
            comments = comment_data.get("comments", [])
            result["fields"]["comment_count"] = comment_data.get("total", len(comments))
            result["fields"]["recent_comments"] = [
                {
                    "author": _user(c.get("author")),
                    "created": c.get("created"),
                    "body": c.get("body"),
                }
                for c in comments[-5:]  # last 5 comments
            ]

        return json.dumps(result, indent=2)
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 404:
            return json.dumps({"error": f"Issue not found: {issue_key}"})
        return json.dumps({"error": f"HTTP {e.response.status_code}: {e.response.text[:200]}"})


async def get_issue_changelog(client: JiraClient, issue_key: str, field: str | None = None) -> str:
    """Get the changelog (edit history) for an issue, optionally filtered to a specific field."""
    try:
        issue = await client.get_issue(issue_key, fields="summary", expand="changelog")
        changelog = issue.get("changelog", {})
        histories = changelog.get("histories", [])

        changes: list[dict] = []
        for h in histories:
            for item in h.get("items", []):
                if field and item.get("field", "").lower() != field.lower():
                    continue
                changes.append({
                    "when": h.get("created"),
                    "who": _user(h.get("author")),
                    "field": item.get("field"),
                    "from": item.get("fromString"),
                    "to": item.get("toString"),
                })

        return json.dumps({
            "key": issue.get("key"),
            "total_history_entries": changelog.get("total", len(histories)),
            "changes": changes,
        }, indent=2)
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 404:
            return json.dumps({"error": f"Issue not found: {issue_key}"})
        return json.dumps({"error": f"HTTP {e.response.status_code}: {e.response.text[:200]}"})


def _name(obj: dict | None) -> str | None:
    return obj.get("name") if obj else None


def _user(obj: dict | None) -> dict | None:
    if not obj:
        return None
    return {
        "key": obj.get("key"),
        "displayName": obj.get("displayName"),
    }

File: src/tools/issues.ts
/** Issue lookup tools. */

import { z } from "zod";
import { isHttpStatusError } from "../errors.js";
import { dumps } from "../json.js";
import type { ToolDef } from "./types.js";
import { has } from "./util.js";

function nameOf(obj: any): string | null {
  return obj ? obj.name : null;
}

function userOf(obj: any): { key: any; displayName: any } | null {
  if (!obj) return null;
  return { key: obj.key, displayName: obj.displayName };
}

export const issueTools: ToolDef[] = [
  {
    name: "get_issue",
    description:
      "Get a Jira issue by key (e.g. 'PROJ-123'). Returns summary, status, type, " +
      "priority, assignee, reporter, labels, components, description, links, and " +
      "recent comments. Optionally pass a comma-separated list of field IDs to restrict output.",
    inputShape: {
      issue_key: z.string().describe("Issue key (e.g. 'PROJ-123')"),
      fields: z
        .string()
        .optional()
        .describe(
          "Comma-separated field IDs to return (e.g. 'summary,status,customfield_10001'). " +
            "Omit for default fields.",
        ),
    },
    async handler({ client }, args) {
      const issueKey = args.issue_key;
      const fields: string | undefined = args.fields;
      try {
        const issue = await client.getIssue(issueKey, fields);

        const result: Record<string, any> = {
          key: issue.key,
          id: issue.id,
          self: issue.self,
        };

        const f = issue.fields ?? {};
        const out: Record<string, any> = {
          summary: f.summary,
          status: nameOf(f.status),
          issuetype: nameOf(f.issuetype),
          priority: nameOf(f.priority),
          resolution: nameOf(f.resolution),
          assignee: userOf(f.assignee),
          reporter: userOf(f.reporter),
          created: f.created,
          updated: f.updated,
          resolutiondate: f.resolutiondate,
          labels: f.labels ?? [],
          components: (f.components ?? []).map((c: any) => c.name),
          description: f.description,
        };
        result.fields = out;

        if (fields) {
          const requested = new Set(fields.split(",").map((fid) => fid.trim()));
          for (const fid of requested) {
            if (fid.startsWith("customfield_") && !has(out, fid)) {
              out[fid] = f[fid];
            }
          }
        }

        const links: any[] = f.issuelinks ?? [];
        if (links.length > 0) {
          out.issuelinks = links.map((link: any) => {
            const linked = link.outwardIssue || link.inwardIssue || {};
            return {
              type: link.type?.name,
              direction: has(link, "outwardIssue") ? "outward" : "inward",
              issue: linked.key,
              summary: (linked.fields ?? {}).summary,
            };
          });
        }

        const commentData = f.comment ?? {};
        if (Object.keys(commentData).length > 0) {
          const comments: any[] = commentData.comments ?? [];
          out.comment_count = has(commentData, "total") ? commentData.total : comments.length;
          out.recent_comments = comments.slice(-5).map((c: any) => ({
            author: userOf(c.author),
            created: c.created,
            body: c.body,
          }));
        }

        return dumps(result);
      } catch (e) {
        if (isHttpStatusError(e)) {
          if (e.status === 404) return dumps({ error: `Issue not found: ${issueKey}` });
          return dumps({ error: `HTTP ${e.status}: ${e.body.slice(0, 200)}` });
        }
        throw e;
      }
    },
  },

  {
    name: "get_issue_changelog",
    description:
      "Get the changelog (edit history) for an issue. Shows who changed what field, " +
      "when, and from/to values. Optionally filter to a single field name " +
      "(e.g. 'assignee', 'status', 'priority'). Use to trace how an issue was modified over time.",
    inputShape: {
      issue_key: z.string().describe("Issue key (e.g. 'PROJ-123')"),
      field: z
        .string()
        .optional()
        .describe(
          "Optional field name to filter changes (e.g. 'assignee', 'status'). " +
            "Omit for all changes.",
        ),
    },
    async handler({ client }, args) {
      const issueKey = args.issue_key;
      const field: string | undefined = args.field;
      try {
        const issue = await client.getIssue(issueKey, "summary", "changelog");
        const changelog = issue.changelog ?? {};
        const histories: any[] = changelog.histories ?? [];

        const changes: any[] = [];
        for (const h of histories) {
          for (const item of h.items ?? []) {
            if (field && (item.field ?? "").toLowerCase() !== field.toLowerCase()) continue;
            changes.push({
              when: h.created,
              who: userOf(h.author),
              field: item.field,
              from: item.fromString,
              to: item.toString,
            });
          }
        }

        return dumps({
          key: issue.key,
          total_history_entries: has(changelog, "total") ? changelog.total : histories.length,
          changes,
        });
      } catch (e) {
        if (isHttpStatusError(e)) {
          if (e.status === 404) return dumps({ error: `Issue not found: ${issueKey}` });
          return dumps({ error: `HTTP ${e.status}: ${e.body.slice(0, 200)}` });
        }
        throw e;
      }
    },
  },
];
